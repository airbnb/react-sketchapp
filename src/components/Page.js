/* @flow */
import React from 'react';
import PropTypes from 'prop-types';
import StyleSheet from '../stylesheet';
import PageStylePropTypes from './PageStylePropTypes';

const propTypes = {
  name: PropTypes.string,
  children: PropTypes.node,
  style: PropTypes.oneOfType([
    PropTypes.shape({ ...PageStylePropTypes }),
    PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.shape({ ...PageStylePropTypes }),
        PropTypes.number,
      ])
    ),
    PropTypes.number,
  ]),
};

class Page extends React.Component {
  static defaultProps = {
    name: 'Page 1',
  };

  render() {
    const { name, children, style, ...otherProps } = this.props;
    const _name =
      name === 'Symbols' ? 'Symbols (renamed to avoid conflict)' : name;
    const _style = StyleSheet.flatten(style);

    return (
      <page name={_name} style={_style} {...otherProps}>
        {children}
      </page>
    );
  }
}

Page.propTypes = propTypes;

module.exports = Page;
