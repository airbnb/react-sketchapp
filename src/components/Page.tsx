import * as React from 'react';
import * as PropTypes from 'prop-types';
import { or } from 'airbnb-prop-types';
import StyleSheet from '../stylesheet';
import PageStylePropTypes from './PageStylePropTypes';

export const PagePropTypes = {
  name: PropTypes.string,
  children: PropTypes.node,
  style: or([PropTypes.shape(PageStylePropTypes), PropTypes.number]),
};

type Props = PropTypes.InferProps<typeof PagePropTypes>;

export default class Page extends React.Component<Props> {
  static propTypes = PagePropTypes;

  render() {
    const { name, children, style, ...otherProps } = this.props;
    const _name = name === 'Symbols' ? 'Symbols (renamed to avoid conflict)' : name;
    const _style = StyleSheet.flatten(style);

    return (
      <sketch_page name={_name} style={_style} {...otherProps}>
        {children}
      </sketch_page>
    );
  }
}
