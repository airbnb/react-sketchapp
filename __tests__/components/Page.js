import React from 'react';
import renderer from 'react-test-renderer';
import Page from '../../src/components/Page';

describe('<Page />', () => {
  it('renders children', () => {
    const tree = renderer
      .create(
        <Page>
          <foo>
            <bar />
          </foo>
        </Page>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('name', () => {
    it('passes its name', () => {
      const tree = renderer.create(<Page name="bar" />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('passes its name and avoids Symbol page conflict', () => {
      const tree = renderer.create(<Page name="Symbols" />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('defaults to Page', () => {
      const tree = renderer.create(<Page />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('passes otherProps', () => {
      const tree = renderer
        .create(<Page propName="something" style={{}} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
