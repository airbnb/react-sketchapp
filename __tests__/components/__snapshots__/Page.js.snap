exports[`<Page /> name defaults to Page 1`] = `
<page
  name="Page 1"
  style={undefined} />
`;

exports[`<Page /> name passes its name 1`] = `
<page
  name="bar"
  style={undefined} />
`;

exports[`<Page /> name passes its name and avoids Symbol page conflict 1`] = `
<page
  name="Symbols (renamed to avoid conflict)"
  style={undefined} />
`;

exports[`<Page /> name passes otherProps 1`] = `
<page
  name="Page 1"
  propName="something"
  style={Object {}} />
`;

exports[`<Page /> renders children 1`] = `
<page
  name="Page 1"
  style={undefined}>
  <foo>
    <bar />
  </foo>
</page>
`;
